lib/pwnlib/asm.rb

```diff
@@ -86,6 +97,60 @@ def asm(code)
       KeystoneEngine::Ks.new(ks_arch, ks_mode).asm(code)[0]
     end
 
+    # Builds an ELF file from executable code.
+    #
+    # @param [String] data
+    #   Assembled code.
+    # @param [Integer?] vma
+    #   The load address for the ELF file.
+    #   If +nil+ is given, default address will be used.
+    #   See {DEFAULT_VMA}.
+    # @param [Boolean] to_file
+    #   Returns ELF content or the path to the ELF file.
+    #   If +true+ is given, the ELF will be saved into a temp file, and the path is returned.
+    #   Otherwise, the content of ELF is returned.
+    #
+    # @return [String]
+    #   If +to_filet+ is +false+ (default), returns the content of ELF. Otherwise, a file is created and the path is
+    #   returned.
+    #
+    # @raise [::Pwnlib::Errors::UnsupportedArchError]
+    #   Raised when don't know how to create an ELF under architecture +context.arch+.
+    #
+    # @todo
+    #   Create a PIE ELF when +vma+ is zero.
+    #
+    # @diff
+    #   Unlike pwntools-python uses cross-compiler to compile code into ELF, we create ELFs in pure Ruby
+    #   implementation, this let us have higher flexibility and less binary dependencies.
+    #
+    # @example
+    #   path = make_elf(asm(shellcraft.cat('/proc/self/maps')), to_file: true)
+    #   puts `#{path}`
+    #   # 08048000-08049000 r-xp 00000000 fd:01 27671233                           /tmp/pwn20180129-3411-7klnng.elf
+    #   # f77c7000-f77c9000 r--p 00000000 00:00 0                                  [vvar]
+    #   # f77c9000-f77cb000 r-xp 00000000 00:00 0                                  [vdso]
+    #   # ffda6000-ffdc8000 rwxp 00000000 00:00 0                                  [stack]
+    def make_elf(data, vma: nil, to_file: false)
```

Comment:
Provide block that accepts path as argument, and remove the file created after yield.